packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,287 @@
+        return [valueKey, unitKey];
+    }
+
+    public cellTemplate: ViewTemplate<ITableCellData<TableColumnTextCellData>> =
+        html<ITableCellData<TableColumnNumberWithUnitCellData>>`
+            <nimble-text-field
+                readonly="true"
+                value=${x => this.formatData(x.data.value, x.data.units)}
+            >
+            </nimble-text-field>
+        `;
+
+    public formatData(value: number, unit: string): string {
+        return `${value.toString()} ${units}`;
+    }
+
+    public validateCellData(cellData: TableRowData): boolean {
+        return typeof cellData['value'] === 'number' && typeof cellData['units'] === 'string';
+    }
+}
+```
+
+Here we have a column visualized in different ways based on custom logic:
+
+```TS
+type TableColumnPositiveNegativeNumberCellData = NumberColumnData<'value'>;
+public TableColumnPositiveNegativeNumber extends FoundationElement implements ITableColumn<TableColumnPositiveNegativeNumberCellData> {
+    ...
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public cellTemplate: ViewTemplate<ITableCellData<TableColumnPositiveNegativeNumberCellData>> =
+        html<ITableCellData<TableColumnPositiveNegativeNumberCellData>`
+            ${when(x => !this.isPositive(x.data.value)),
+            html`
+                <nimble-text-field
+                    readonly="true"
+                    value=${x => x.data.value}
+                    style="fontColor: red;"
+                >
+                </nimble-text-field>
+            `}
+            ${when(x => this.isPositive(x.data.value)),
+            html`
+                <nimble-text-field
+                    readonly="true"
+                    value=${x => x.data.value}
+                    style="fontColor: green;"
+                >
+                </nimble-text-field>
+            `}
+        `;
+
+    public isPositive(value: number): bool {
+        return value >= 0;
+    }
+
+    public validateCellData(cellData: TableRowData): boolean {
+        return typeof cellData['value'] === 'number';
+    }
+}
+```
+
+Finally, here is a column that allows a user to register a callback for a click event on a button inside the cell template:
+
+```TS
+type TableColumnButtonCellData = StringColumnData<'id'>;
+public TableColumnButton extends FoundationElement implements ITableColumn<TableColumnButtonCellData> {
+    ...
+
+    public cellDataKeyNames = ['id'] as const;
+
+    @attr
+    public idKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public callback: (id: string) => void;
+
+    public cellTemplate: ViewTemplate<ITableCellData<TableColumnButtonCellData>> =
+        html<ITableCellData<TableColumnButtonCellData>>`
+            <nimble-button readonly="true" @click="${x => this.callback(x.data.id)}>
```

Comment:
`readonly="true"` doesn't mean anything for a button, right?

Reply:
Copy-pasta.